File: map.cpp
#include <iostream>
#include <fstream>
#include <cstring>
#include <string>
#include <sstream>
#include <iomanip>
#include <Windows.h>
using namespace std;

short readShort(char* str) {
    char* buf = new char[2];
    memcpy(buf, str, 2);
    short *p = (short *)buf;
    return *p;
}

int readInt(char* str) {
    char* buf = new char[4];
    memcpy(buf, str, 4);
    int *p = (int *)buf;
    return *p;
}

float readFloat(char* str) {
    char* buf = new char[4];
    memcpy(buf, str, 4);
    float *p = (float *)buf;
    return *p;
}

string readString(char* str, int length) {
    char* buf = new char[length];
    memcpy(buf, str, length);
    buf[length] = '\0';

    return string(buf, length);
}

int main() {
    SetConsoleOutputCP(CP_UTF8);

    string sbuf;
    string input;
    cout << "DEND MAP SCRIPT ver1.0.0..." << endl;
    cout << "railのbinファイル名を入力してください: ";
    cin >> input;

    ifstream fin(input, ios::in | ios::binary);
    if (!fin) {
        cin.ignore();
        cout << "指定されたファイルが見つかりません。終了します。" << endl;
        getchar();
        return -1;
    }

    printf("見つけました！\n\n");
    fin.seekg(0, ios::end);
    int size = fin.tellg();
    fin.seekg(0, ios::beg);
    char* str = new char[size];
    fin.read(str, size);

    fin.close();

    int index = 16;
    bool readFlag = false;
    sbuf = readString(&str[0], index);

    if (sbuf != "DEND_MAP_VER0300" && sbuf != "DEND_MAP_VER0400") {
        cin.ignore();
        cout << "DEND map scriptファイルではありません。終了します。" << endl;
        cout << sbuf << endl;
        getchar();
        return -1;
    }

    if (sbuf == "DEND_MAP_VER0400") {
        readFlag = true;
    }

    int musicCnt = str[index];
    index++;
    for (int i = 0; i < musicCnt; i++) {
        index++;
    }

    int pRailCnt = str[index];
    index++;
    for (int i = 0; i < pRailCnt; i++) {
        for (int k = 0; k < 4; k++) {
            index += 2;
        }
        printf("\n");
        index++;
    }

    // unknown(1)
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 4; j++) {
            index += 2;
        }
        index++;
    }
    // unknown(1)
    
    index++;
    // 描画量? (1700)
    index += 4;
    
    //unknown(2)
    int unknownCnt = str[index];
    index++;
    for (int i = 0; i < unknownCnt; i++) {
        for (int j = 0; j < 2; j++) {
            index += 4;
        }
        index += 3;
    }
    //unknown(2)

    cout << "Read Light..." << endl;
    int lightCnt = str[index];
    index++;
    for (int i = 0; i < lightCnt; i++) {
        int b = str[index];
        index++;
        cout << i << " -> " << readString(&str[index], b) << endl;
        index += b;
    }
    cout << endl;

    cout << "Read png..." << endl;
    short pngCnt = readShort(&str[index]);
    index += 2;
    for (int i = 0; i < pngCnt; i++) {
        int b = str[index];
        index++;
        cout << i << " -> " << readString(&str[index], b) << endl;
        index += b;
    }
    cout << endl;

    //unknown(3)
    short cnt = readShort(&str[index]);
    index += 2;
    for (int i = 0; i < cnt; i++) {
        index += 9;
    }
    //unknown(3)

    cout << "Read Object Bin..." << endl;
    int binCnt = str[index];
    index++;
    for (int i = 0; i < binCnt; i++) {
        int b = str[index];
        index++;
        cout << i << " -> " << readString(&str[index], b) << endl;
        index += b;
    }
    cout << endl;
    index += 4;

    index += 2;
    cout << "Read smf..." << endl;
    char* buf = new char[2];
    memcpy(buf, &str[index], 1);
    buf[1] = 0x00;
    delete buf;
    short* smfCnt = (short *)buf;
    index++;
    for (int i = 0; i < *smfCnt; i++) {
        int b = str[index];
        index++;
        cout << i << " -> " << readString(&str[index], b);
        index += b;

        cout << " [";
        for (int j = 0; j < 4; j++) {
            short num = readShort(&str[index]);
            cout << num << ", ";
            index += 2;
        }
        cout << "]" << endl;
    }
    cout << endl;

    cout << "Read Station Name..." << endl;

    int snameCnt = str[index];
    index++;

    SetConsoleOutputCP(932);
    for (int i = 0; i < snameCnt; i++) {
        int b = str[index];
        index++;
        cout << i << " -> [ ";
        if (b > 0) {
            string buf = readString(&str[index], b);
            index += b;

            cout << buf;
        }
        cout << " ] ";
        for (int j = 0; j < 3; j++) {
            cout << (int)str[index++] << ", ";
        }
        cout << endl;
        index += 26;
    }
    cout << endl;

    SetConsoleOutputCP(CP_UTF8);

    //unknown(4)
    if (str[index] == 0x01) {
        index += 0x12;
    } else {
        index++;
    }
    //unknown(4)

    cout << "Read cpu data..." << endl;
    int cpuCnt = str[index];
    index++;

    for (int i = 0; i < cpuCnt; i++) {
        cout << i << " -> ";
        //Rail.No
        short railNo = readShort(&str[index]);
        cout << railNo << ", ";
        index += 2;
        //CPU mode?
        cout << (int)str[index++] << ", ";
        cout << (int)str[index++] << ", ";
        //
        for (int j = 0; j < 4; j++) {
            float f = readFloat(&str[index]);
            cout << f << ", ";
            index += 4;
        }
        cout << endl;
    }
    cout << endl;

    cout << "Read comic bin data..." << endl;
    int comicbinCnt = str[index];
    index++;

    for (int i = 0; i < comicbinCnt; i++) {
        cout << i << " -> ";
        short comicName = readShort(&str[index]);
        cout << comicName << ", ";
        index += 2;
        cout << (int)str[index++] << ", ";
        cout << (int)str[index++] << ", ";
        cout << (int)str[index++] << ", " << endl;
    }
    cout << endl;

    cout << "Dosan data...?" << endl;
    int dosanCnt = str[index];
    index++;

    for (int i = 0; i < dosanCnt; i++) {
        cout << i << " -> ";
        cout << "s_rail : [";
        for (int j = 0; j < 3; j++) {
            short num = readShort(&str[index]);
            cout << num << ", ";
            index += 2;
        }
        cout << "], " << endl;

        cout << "    e_rail : [";
        for (int j = 0; j < 3; j++) {
            short num = readShort(&str[index]);
            cout << num << ", ";
            index += 2;
        }
        cout << "], " << endl;
        
        cout << "    offset : ";
        short num = readShort(&str[index]);
        cout << num << endl;
        index += 2;

        cout << "    ";
        for (int j = 0; j < 5; j++) {
            float f = readFloat(&str[index]);
            cout << f << ", ";
            index += 4;
        }

        num = readShort(&str[index]);
        cout << num << endl;
        index += 2;
    }
    cout << endl;

    cout << hex << index << dec << endl;
    cout << "Read Map Data..." << endl;
    bool flag;
    bool onlyDrift = true;
    short mapCnt = readShort(&str[index]);
    index += 2;
    for (int i = 0; i < mapCnt; i++) {
        flag = false;
        ostringstream oss;
        oss << "[";
        short rail_no = readShort(&str[index]);
        index += 2;
        oss << rail_no << ", ";
        char block = str[index];
        index++;
        oss << (int)block;
        oss << "]";

        oss << ", ";
        //vector
        oss << "[";
        for (int j = 0; j < 3; j++) {
            float xyz = readFloat(&str[index]);;
            index += 4;
            oss << xyz << ", ";
        }
        oss << "], ";
        oss << "[";
        short mdl_no = readShort(&str[index]);
        oss << mdl_no << ", ";
        index += 2;
        oss << (int)str[index] << "], ";
        index++;

        float per = readFloat(&str[index]);;
        oss << per << ", ";
        index += 4;

        int k = 0;
        char* flg = new char[4];
        memcpy(flg, &str[index], 4);
        index += 4;
        
        for (int j = 0; j < 4; j++) {
            oss << "0x";
            oss << hex << setfill('0') << setw(2) << (flg[k] & 0x000000FF) << dec << ", ";
            if (j == 3) {
                if ((flg[k] & 0x01) == 0x01) {
                    flag = true;
                }
                if ((flg[k] & 0x02) == 0x02) {
                    flag = true;
                }
                if ((flg[k] & 0x04) == 0x04) {
                    flag = true;
                }
            }
            k++;
        }
        delete flg;

        int rail_data = str[index];
        index++;

        for (int j = 0; j < rail_data; j++) {
            if (readFlag) {
                oss << "[";
                short next_rail = readShort(&str[index]);
                index += 2;
                oss << next_rail << ", ";
                short next_no = readShort(&str[index]);
                index += 2;
                oss << next_no << ", ";
                short prev_rail = readShort(&str[index]);
                index += 2;
                oss << prev_rail << ", ";
                short prev_no = readShort(&str[index]);
                index += 2;
                oss << prev_no << "], ";
            }
            oss << "[";
            short next_rail = readShort(&str[index]);
            index += 2;
            oss << next_rail << ", ";
            short next_no = readShort(&str[index]);
            index += 2;
            oss << next_no << ", ";
            short prev_rail = readShort(&str[index]);
            index += 2;
            oss << prev_rail << ", ";
            short prev_no = readShort(&str[index]);
            index += 2;
            oss << prev_no << "], ";
        }
        oss << endl;
        cout << oss.str();
        /*
        delete buf;
        if (!onlyDrift) {
            string str = oss.str();
            cout << str;
        } else {
            if (flag) {
                string str = oss.str();
                cout << str;
            }
        }
        */
    }

    /*
    cout << "正常に読み込みできました。終了します。" << endl;
    if (index < size) {
        cout << "注意！設定したコマンド数(" << count << ")は、書き込んだコマンド数より少なく設定されています" << endl;
    }
    getchar();
    */

    return 0;
}

File: map.py
import struct
import os
import sys

def readBinary(line, mode):
    if mode == "short":
        s = struct.unpack("<h", line)[0]
        return s
    elif mode == "ushort":
        s = struct.unpack("<H", line)[0]
        return s
    elif mode == "int":
        i = struct.unpack("<i", line)[0]
        return i
    elif mode == "float":
        f = struct.unpack("<f", line)[0]
        return f
    else:
        return None

print("DEND MAP SCRIPT ver1.0.0...")
file = input("railのbinファイル名を入力してください: ")

readFlag = False

try:
    try:
        f = open(file, "rb")
        line = f.read()
        f.close()
    except FileNotFoundError:
        errorMsg = "指定されたファイルが見つかりません。終了します。"
        print(errorMsg)
        input()
        sys.exit()

    print("見つけました！")
    size = len(line)

    index = 16
    header = line[0:index]
    if header != b'DEND_MAP_VER0300' and header != b'DEND_MAP_VER0400':
        raise Exception

    if header == b'DEND_MAP_VER0400':
        readFlag = True

    #使う音楽(ダミーデータ?)
    musicCnt = line[index]
    index += 1
    for i in range(musicCnt):
        index += 1

    #配置する車両カウント
    trainCnt = line[index]
    print("初期配置カウント：{0}".format(trainCnt))
    index += 1
    print()

    print("初期位置")
    #3車両の初期レール位置
    for i in range(3):
        railNo = readBinary(line[index:index+2], "short")
        index += 2
        boneNo = readBinary(line[index:index+2], "short")
        index += 2

        ##########unknown
        index += 4
        index += 1
        ##########unknown
        print("{0} -> ({1}, {2})".format(i+1, railNo, boneNo))
    print()

    print("試運転位置")
    #試運転の初期レール位置
    railNo = readBinary(line[index:index+2], "short")
    index += 2
    boneNo = readBinary(line[index:index+2], "short")
    index += 2

    ##########unknown
    index += 4
    index += 1
    ##########unknown
    print("({0}, {1})".format(railNo, boneNo))
    print()

    #二人バトルの初期レール位置
    print("二人バトル位置")
    for i in range(2):
        railNo = readBinary(line[index:index+2], "short")
        index += 2
        boneNo = readBinary(line[index:index+2], "short")
        index += 2

        ##########unknown
        index += 4
        index += 1
        ##########unknown
        print("{0} -> ({1}, {2})".format(i+1, railNo, boneNo))
    print()

    index += 1

    ##########unknown
    print(readBinary(line[index:index+4], "float"))
    index += 4
    print()

    cnt = line[index]
    index += 1
    for i in range(cnt):
        for j in range(2):
            print(readBinary(line[index:index+4], "float"))
            index += 4
        index += 3
    print()
    ##########unknown
    
    #Light情報
    print("Read Light...")
    lightCnt = line[index]
    index += 1
    for i in range(lightCnt):
        b = line[index]
        index += 1
        text = line[index:index+b].decode()
        print("{0} -> {1}".format(i, text))
        index += b
    print()

    #png情報
    print("Read Png...")
    pngCnt = readBinary(line[index:index+2], "short")
    index += 2
    for i in range(pngCnt):
        b = line[index]
        index += 1
        text = line[index:index+b].decode()
        print("{0} -> {1}".format(i, text))
        index += b
    print()

    ##########unknown
    cnt = readBinary(line[index:index+2], "short")
    index += 2
    for i in range(cnt):
        index += 9
    ##########unknown

    print("Read Object Bin...")
    binCnt = line[index]
    index += 1
    for i in range(binCnt):
        b = line[index]
        index += 1
        text = line[index:index+b].decode()
        print("{0} -> {1}".format(i, text))
        index += b
    print()

    ##########unknown
    cnt = line[index]
    index += 1
    for i in range(cnt):
        index += 5
    ##########unknown
    
    print("Read smf...")
    smfCnt = line[index]
    index += 1
    for i in range(smfCnt):
        b = line[index]
        index += 1
        text = line[index:index+b].decode()
        index += b
        arr = []
        for j in range(4):
            n = readBinary(line[index:index+2], "short")
            arr.append(n)
            index += 2
        print("{0} -> {1}, {2}".format(i, text, arr))
    print()

    print("Read Station Name...")
    snameCnt = line[index]
    index += 1
    for i in range(snameCnt-1):
        b = line[index]
        index += 1
        text = ""
        if b > 0:
            text = line[index:index+b].decode("shift-jis")
            index += b
        arr = []
        for j in range(3):
            arr.append(line[index])
            index += 1
        index += 0x1A
        print("{0} -> {1}, {2}".format(i, text, arr))

    index += 1

    ##########unknown
    cnt = line[index]
    index += 1
    for i in range(cnt):
        index += 0x0E

    cnt = line[index]
    index += 1
    if cnt > 0:
        for i in range(cnt):
            index += 16

        index += 1
    print()
    ##########unknown
    
    #cpu
    print("Read cpu data...")
    cpuCnt = line[index]
    index += 1
    for i in range(cpuCnt):
        arr = []
        railNo = readBinary(line[index:index+2], "short")
        index += 2
        arr.append(railNo)
        
        org = line[index]
        index += 1
        arr.append(org)
        mode = line[index]
        index += 1
        arr.append(mode)

        for j in range(4):
            f = readBinary(line[index:index+4], "float")
            index += 4
            arr.append(f)
        print("{0} -> {1}".format(i, arr))
    print()

    #comic bin data
    print("Read comic bin data...")
    comicbinCnt = line[index]
    index += 1

    for i in range(comicbinCnt):
        arr = []
        comicName = readBinary(line[index:index+2], "short")
        index += 2
        arr.append(comicName)

        for j in range(3):
            arr.append(line[index])
            index += 1
        print("{0} -> {1}".format(i, arr))
    print()

    #Dosan data...?
    dosanCnt = line[index]
    index += 1

    for i in range(dosanCnt):
        print("{0} -> ".format(i))
        arr = []
        for j in range(3):
            num = readBinary(line[index:index+2], "short")
            index += 2
            arr.append(num)
        print("s_rail：{0}".format(arr))

        arr = []
        for j in range(3):
            num = readBinary(line[index:index+2], "short")
            index += 2
            arr.append(num)
        print("e_rail：{0}".format(arr))

        num = readBinary(line[index:index+2], "short")
        index += 2
        print("offset：{0}".format(num))

        arr = []
        for j in range(5):
            f = readBinary(line[index:index+4], "float")
            index += 4
            arr.append(f)
        print("float：{0}".format(arr))

        num = readBinary(line[index:index+2], "short")
        index += 2
        print("short：{0}".format(num))
    print()

    #Map
    print("Read Map Data...")
    mapCnt = readBinary(line[index:index+2], "short")
    index += 2

    for i in range(mapCnt):
        rail_no = readBinary(line[index:index+2], "short")
        index += 2
        block = line[index]
        index += 1

        print("{0} -> [{1},{2}], ".format(i, rail_no, block), end="")

        #vector?
        xyz = []
        for j in range(3):
            f = readBinary(line[index:index+4], "float")
            xyz.append(f)
            index += 4
        print("{0}, ".format(xyz), end="")

        mdl_no = line[index]
        index += 1

        mdl_no_next = readBinary(line[index:index+2], "short")
        index += 2

        per = readBinary(line[index:index+4], "float")
        index += 4

        print("[{0},{1}], {2}, ".format(mdl_no, mdl_no_next, per), end="")
        
        #flg
        flg = []
        for j in range(4):
            flg.append(hex(line[index]))
            index += 1
        print("{0}, ".format(flg), end="")

        #rail_data
        rail_data = line[index]
        index += 1

        for j in range(rail_data):
            if readFlag:
                next_rail = readBinary(line[index:index+2], "short")
                index += 2
                next_no = readBinary(line[index:index+2], "short")
                index += 2
                prev_rail = readBinary(line[index:index+2], "short")
                index += 2
                prev_no = readBinary(line[index:index+2], "short")
                index += 2

                print("[{0},{1},{2},{3}], ".format(next_rail, next_no, prev_rail, prev_no), end="")

            next_rail = readBinary(line[index:index+2], "short")
            index += 2
            next_no = readBinary(line[index:index+2], "short")
            index += 2
            prev_rail = readBinary(line[index:index+2], "short")
            index += 2
            prev_no = readBinary(line[index:index+2], "short")
            index += 2
            print("[{0},{1},{2},{3}], ".format(next_rail, next_no, prev_rail, prev_no), end="")

        print()
    input()
        
except Exception as e:
    print(e)
    sys.exit()


